src/modules/auth/auth.service.ts

```diff
@@ -34,5 +43,126 @@ export class AuthService {
+            createdAt: true,
+            updatedAt: true,
+            phoneNumber: true,
+            password: true,
+            address: true,
+            name: true,
+            gender: true,
+            role: true,
+            birthDate: true,
+            status: true,
+            image: true,
+            refreshToken: true,},
+        });
+    
+        if (!user) {
+          throw new HttpException('User does not exist', HttpStatus.UNAUTHORIZED);
+        }
+    
+        const isPasswordValid = await bcrypt.compare(loginDto.password, user.password);
+    
+        if (!isPasswordValid) {
+          throw new UnauthorizedException('Invalid credentials');
+        }
+    
+        const refreshToken = await this.generateRefreshToken(user);
+        
+        await this.prismaService.users.update({
+          where: { id: user.id },
+          data: { refreshToken },
+        });
+    
+        const accessToken = await this.generateAccessToken(user);
+    
+        return { accessToken , refreshToken };
+      }
+    
+      async logout(refreshToken: string) {
+        const decodedToken = this.jwtService.decode(refreshToken) as { id: number };
+        const user = await this.prismaService.users.findUnique({
+          where: { id: decodedToken.id },
+          select: { refreshToken: true },
+        });
+      
+        if (user && user.refreshToken) {
+          return this.prismaService.users.update({
+            where: { id: decodedToken.id },
+            data: { refreshToken: null },
+          });
+        } else {
+          throw new UnauthorizedException('Invalid refreshToken');
+        }
+      }
+
+
+    // auth.service.ts
+    async refreshToken(refreshToken: string) {
+      try {
+        const decodedToken = this.jwtService.decode(refreshToken) as { id: number };
+        const user = await this.prismaService.users.findUnique({
+          where: { id: decodedToken.id },
+          select: {
+            id: true,
+            createdAt: true,
+            updatedAt: true,
+            phoneNumber: true,
+            password: true,
+            address: true,
+            name: true,
+            gender: true,
+            role: true,
+            birthDate: true,
+            status: true,
+            image: true,
+            refreshToken: true,
+          },
+        });
+
+        if (user && user.refreshToken === refreshToken && !this.isTokenExpired(refreshToken)) {
+          const newAccessToken = await this.generateAccessToken(user);
+          return { accessToken: newAccessToken };
+        } else {
+          throw new UnauthorizedException('Invalid or expired refreshToken');
+        }
+      } catch (error) {
+        // Handle token decoding or database lookup errors
+        throw new UnauthorizedException('Invalid refreshToken');
+      }
+    }
+
+    private isTokenExpired(token: string): boolean {
+      const decodedToken = this.jwtService.decode(token, { json: true }) as { exp: number };
```

Comment:
Create a Type for your Payload, then use it for return type here. Dont cast by object on every decode.